src/main/java/org/folio/auth/authtokenmodule/tokens/Token.java

```diff
@@ -0,0 +1,249 @@
+    // it is considered to be encrypted. If there are not 5 it is rejected from JWE parsing.
+    // Our unencrypted tokens have 3 parts (and 2 separators).
+    int parts = token.split("\\.").length;
+    if (parts == 5)
+      return true;
+
+    if (parts == 3)
+      return false;
+
+    throw new TokenValidationException("Unexpected token part count", 401);
+  }
+
+  /**
+   * Validate all the things that tokens have in common.
+   * @param request The http request context where the token is being provided.
+   * @throws TokenValidationException Throws this when any validation step fails, but may throw
+   * other exceptions as well.
+   */
+  protected void validateCommon(HttpServerRequest request) throws TokenValidationException {
+    try {
+      // Check that the token has a source.
+      if (source == null)
+        throw new TokenValidationException("Token has no source defined", 500);
+
+      // Check that the claims have been created.
+      if (claims == null)
+        throw new TokenValidationException("Token has no claims", 500);
+
+      // Check some claims that all tokens must have.
+      String[] requiredClaims = new String[] { "sub", "tenant", "type" };
+      for (String c : requiredClaims) {
+        if (!claims.containsKey(c)) {
+          throw new TokenValidationException(String.format("Token is missing %s claim", c), 500);
+        }
+      }
+
+      if (request == null)
+        return;
+
+      // Check that some items in the headers match what are in the token.
+      String headerTenant = request.headers().get(XOkapiHeaders.TENANT);
+      if (!claims.getString("tenant").equals(headerTenant))
+        throw new TokenValidationException("Tenant in header does not equal tenant in token", 403);
+
+      String headerUserId = request.headers().get(XOkapiHeaders.USER_ID);
+      if (headerUserId != null && !claims.getString("user_id").equals(headerUserId))
+        throw new TokenValidationException("User id in header does not equal user id in token", 403);
+
+    } catch (TokenValidationException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new TokenValidationException("Unexpected token validation exception", e, 500);
+    }
+  }
+
+  private static Token parse(String sourceToken, TokenCreator tokenCreator) throws TokenValidationException {
+    Token token = null;
+    JsonObject claims = null;
+    final String invalidTokenMsg = "Invalid token";
+
+    try {
+      if (isEncrypted(sourceToken)) {
+        String tokenContent = tokenCreator.decodeJWEToken(sourceToken);
+        claims = new JsonObject(tokenContent);
+      } else {
+        tokenCreator.checkJWTToken(sourceToken);
+        claims = getClaims(sourceToken);
+      }
+    } catch (ParseException p) {
+      throw new TokenValidationException(invalidTokenMsg, p, 401);
+    } catch (JOSEException j) {
+      throw new TokenValidationException(invalidTokenMsg, j, 401);
```

Comment:
You can merge these two catch statements

Reply:
I'm not sure these should be combined. `checkJWToken` throws all three of these exceptions and I think it is desirable to pass these along to the `TokenValidationException` which allows for each to be logged in the module's error logger. But maybe I'm misunderstanding.